tidy3d/components/source.py

```diff
@@ -291,7 +294,81 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence, real or complex valued.
+
+    Note
+    ----
+    The source time dependence is linearly interpolated to the simulation time steps.
+    To ensure that this interpolation does not introduce artifacts, it is necessary
+    to use a sampling rate that is sufficiently fast relative to the simulation time step.
+
+    Example
+    -------
+    >>> cst = td.CustomSourceTime.from_values(freq0=1, fwidth=0.1,
+    >>>     values=np.linspace(0, 9, 10), dt=0.1)
+    """
+
+    source_time_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _source_time_dataset_none_warning = warn_if_dataset_none("source_time_dataset")
+
+    @classmethod
+    def from_values(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> CustomSourceTime:
+        """Create a :class:`.CustomSourceTime` from a numpy array.
+
+        Parameters
+        ----------
+        freq0 : float
+            Estimated central frequency of the source.
+        fwidth : float
+            Estimated frequency width of the source.
+        values: ArrayComplex1D
+            Complex values of the source amplitude.
+        dt: float
+            Time step for the `values` array. This value should be sufficiently small
+            relative to the simulation `dt` in order to avoid interpolation artifacts.
+
+
+        Returns
+        -------
+        CustomSourceTime
+            :class:`.CustomSourceTime` with these values, and time coordinates evenly spaced
+            between 0 and dt * (N-1) with a step size of `dt`, where N is the length of
+            the values array.
+        """
+        times = np.arange(len(values)) * dt
+        source_time_dataarray = TimeDataArray(values, coords=dict(t=times))
+        source_time_dataset = TimeDataset(values=source_time_dataarray)
+        return CustomSourceTime(
+            freq0=freq0,
+            fwidth=fwidth,
+            source_time_dataset=source_time_dataset,
+        )
+
+    def amp_time(self, time: float) -> complex:
+        """Complex-valued source amplitude as a function of time.
+
+        Parameters
+        ----------
+        time : float
+            Time in seconds.
+
+        Returns
+        -------
+        complex
+            Complex-valued source amplitude at that time.
+        """
+        if self.source_time_dataset is None:
+            return None
+        return self.source_time_dataset.values.interp(t=time).to_numpy()
```

Comment:
if we are going with `interp`, there are a couple edge cases brought on by xarray's API that should probably be addressed. 
1. if the source_time_dataset only has a single time value, it will raise a very obscure message.
2. if `time` is out of range of the dataset, I believe it will return nans, so we either need to 
    a. convert nans to 0 (`np.nan_to_num`)
    b. use an interpolation scheme where out of range values use the closest data point.
    c. validate (in the Simulation-level validator) that the dataset covers all of the possible `Simulation.tmesh`

So I think it would be good to

Reply:
Is the 2nd issue resolved in the current status?

Reply:
I added a Simulation-level validator that it covers the Simulation.tmesh, as well as a test for this error